### PR TITLE
Updates to make features based on atomic ops optional, to support thumbv6m-none-eabi

### PR DESCRIPTION
**Pull request by apullin**
These modifications appear to build and work as expected on thumbv6 (without atomics) and thumbv7 (with atomics) system using the nightly toolchain.

Tests pass on x64_64 with both stable and nightly toolchains.

I am not quite sure if the arrangement for the added `nightly` feature is not the correct way to gate the needed attributes since this means that functionality would be lost going from 0.1.2 -> 0.1.3 (this version) when using stable.

I am also not sure if the caution around keeping the vestigial `alloc` feature is needed, or if it should be dropped entirely.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,4 +14,7 @@ cty = "0.1"
 memchr = { version = "2.0", default-features = false }
 
 [features]
+# This feature no longer does anything, because alloc is stable.
+# It is still included here to not break builds depending on this crate.
 alloc = []
```
Comment (Amanieu) [inline Cargo.toml:18]: Just remove the alloc feature, I'm going to do a major version bump.

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "cstr_core"
-version = "0.1.2"
+version = "0.1.3"
```
Comment (Amanieu) [inline Cargo.toml]: I will take care of bumping the version number separately when I publish the crate.

File: Cargo.toml [hunk 3]
```diff
@@ -14,4 +14,7 @@ cty = "0.1"
 memchr = { version = "2.0", default-features = false }
 
 [features]
+# This feature no longer does anything, because alloc is stable.
+# It is still included here to not break builds depending on this crate.
 alloc = []
+nightly = []
```
Comment (Amanieu) [inline Cargo.toml]: Make sure the nightly feature gets tested in travis.

File: src/lib.rs [hunk 4]
```diff
@@ -9,30 +9,25 @@
...
 extern crate std;
 
-#[cfg(feature = "alloc")]
 extern crate alloc;
 extern crate cty;
 extern crate memchr;
 
-#[cfg(feature = "alloc")]
+#[cfg(feature = "nightly")]
+#[cfg(target_has_atomic = "ptr")]
```
Comment (Amanieu) [inline src/lib.rs]: I don't think this condition does what you describe. When the nightly feature is disabled, we want to enable `Arc` by default.

Try something along the lines of:
`#[cfg_attr(feature = "nightly", cfg(target_has_atomic = "ptr")]`

**Review by Amanieu — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 4 above)
Comment (apullin) [reply src/lib.rs]: This is tough. I was trying to come up with a solution that would allow both:
- building with stable and inclusion of Arc without the nightly-specific atomic check
- building with nightly with the 'nightly' feature to check atomic checks

The `cfg` setup is a little elaborate, and still exposes the failure case the `nightly` feature is specified when using the stable toolchain. But maybe that is OK.

File: Cargo.toml [hunk 5]
```diff
@@ -12,7 +12,8 @@ keywords = ["cstr", "cstring", "no_std", "string", "c"]
 [dependencies]
 cty = "0.2.1"
 memchr = { version = "2.2.1", default-features = false }
+rustversion = "1.0.2"
```
Comment (Amanieu) [inline Cargo.toml]: This dependency isn't used.

File: src/lib.rs [hunk 6]
```diff
@@ -9,30 +9,33 @@
...
-#[cfg(feature = "alloc")]
+// If building with the nightly toolchain and "nightly" feature, we can optionally
+// check if our target actually supports atomic operations.
+// If building with stable, we will include Arc, but the build may fail here for some targets.
+#[cfg(any(
+        not(feature="nightly"),
+        all(feature = "nightly", target_has_atomic = "ptr")
+         )
+     )
+]
```
Comment (Amanieu) [inline src/lib.rs]: Actually I think there is a better way to handle this: just add an `arc` feature that is enabled by default. That way users on thumbv6 targets can explicitly disable the feature.

File: src/lib.rs [hunk 7]
```diff
@@ -9,30 +9,33 @@
...
 #![no_std]
-#![cfg_attr(feature = "alloc", feature(alloc))]
+
+#![cfg_attr(feature = "nightly", feature(cfg_target_has_atomic))]
 
 #[cfg(test)]
 #[macro_use]
 extern crate std;
 
-#[cfg(feature = "alloc")]
```
Comment (Amanieu) [inline src/lib.rs]: I missed this in the initial review but could you revert these changes? There are good reasons for wanting to avoid a dependency on the `alloc` crate, such as systems that don't have a global allocator. It's fine if the `alloc` feature is enabled by default, as long as there is a way to disable it.

**Review by Amanieu — CHANGES_REQUESTED** [review]

**Comment by Amanieu** [comment]
Thanks!

**Comment by Amanieu** [comment]
Published version 0.2.0 of the crate.
